tests/testsuite/test.rs

```diff
@@ -3658,3 +3658,160 @@ fn test_dep_with_dev() {
         )
         .run();
 }
+
+#[cfg(target_env = "gnu")]
+#[cargo_test]
+fn cargo_test_doctest_xcompile_ignores() {
+    use crate::support::is_nightly;
+    if !is_nightly() {
+        return;
+    }
+    let p = project()
+        .file("Cargo.toml", &basic_lib_manifest("foo"))
+        .file(
+            "src/lib.rs",
+            r#"
+
+            ///```ignore-gnu
+            ///assert!(1 == 1);
+            ///```
+            pub fn foo() -> u8 {
+                4
+            }
+            "#,
+        )
+        .build();
+
+    p.cargo("build").run();
+    p.cargo("test")
+        .with_stdout_contains(
+            "\
+             test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out\
+             ",
+        )
+        .run();
+    p.cargo("test -Zdoctest-xcompile")
+        .masquerade_as_nightly_cargo()
+        .with_stdout_contains(
+            "\
+             test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out\
+             ",
+        )
+        .run();
+}
+
+#[cfg(target_env = "gnu")]
+#[cargo_test]
+fn cargo_test_doctest_xcompile() {
+    use crate::support::is_nightly;
+    if !is_nightly() {
+        return;
+    }
+    let p = project()
+        .file("Cargo.toml", &basic_lib_manifest("foo"))
+        .file(
+            "src/lib.rs",
+            r#"
+
+            ///```
+            ///assert!(1 == 1);
+            ///```
+            pub fn foo() -> u8 {
+                4
+            }
+            "#,
+        )
+        .build();
+
+    p.cargo("build").run();
+    p.cargo("test --target i686-unknown-linux-gnu")
+        .with_stdout_contains(
+            "\
+             running 0 tests\
+             ",
+        )
+        .run();
+    p.cargo("test --target i686-unknown-linux-gnu -Zdoctest-xcompile")
+        .masquerade_as_nightly_cargo()
+        .with_stdout_contains(
+            "\
+             test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out\
+             ",
+        )
+        .run();
+}
+
+#[cfg(target_env = "gnu")]
```

Comment:
I don't think there's any hard requirement that this is a gnu-only test, could this use Cargo to produce an executable so this test can be run on all platforms? (there's examples of how to do this in other tests)